Internode gpu fallback for OFI

gcongiu opened the pull request:
## Pull Request Description

This PR provides inter-node fallback path for GPU data in OFI direct netmod in pt2pt communication. For send operations, whenever OFI detects a GPU buffer it allocates a host registered stage buffer into which GPU data is moved, doing packing whenever necessary (i.e. user specified non-contiguous datatype). The stage buffer is then passed down to libfabric for transfer. The allocated stage buffer is tracked by the MPI request and is freed upon request completion. For receiver operations, whenever OFI detects a GPU buffer it allocates a host registered stage buffer. A reference to the buffer, as well as the datatype and the source GPU buffer, are cached in the MPI request. The stage buffer is then used to receive contiguous data from the network. When the request is completed the datatype information in the request is used to move the data from the stage buffer to the source GPU buffer, doing unpacking whenever necessary (i.e. user specified non-contiguous datatype). Afterwards, the stage buffer is freed.
 
<!--
Insert description of the work in this merge request (above this comment),
particularly focused on _why_ the work is necessary, not _what_ you did.
-->

<!-- AUTHOR: After creating this merge request, check off each of the following items as you complete them. -->

## Expected Impact

## Author Checklist
* [ ] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [ ] Remove xfail from the test suite when fixing a test
* [ ] Commits are self-contained and do not do two things at once
* [ ] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [ ] Passes whitespace checkers
* [ ] Passes warning tests
* [ ] Passes all tests
* [ ] Add comments such that someone without knowledge of the code could understand
* [ ] Add Devel Docs in the `doc/` directory for any new code design

gcongiu commented:
test:jenkins/ch4/ofi

gcongiu commented:
@yfguo although this is not ready to review yet, would you mind taking a quick look at the overall structure of the framework and see if there is any major problem with it?

gcongiu commented on src/mpid/ch4/netmod/ofi/ofi_send.h:
```diff
@@ -480,6 +480,12 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_NM_mpi_send(const void *buf, MPI_Aint count,
+#ifdef HAVE_GPU
+    if (MPIDI_IS_BUF_TYPE_GPU(buf) && !MPIDI_OFI_global.can_handle_gpu_mem) {
+        mpi_errno =
+            MPIDIG_gpu_send(buf, count, datatype, rank, tag, comm, context_offset, addr, request);
```
@hzhou probably the GPU fallback code should be moved to `mpidig`. If I am not mistaken, this is intended to handle whatever case the netmod cannot handle (e.g., gpu memory). However, I am not sure this is appropriate as `mpidig` is currently used for active messages only. For the GPU fallback we don't need to use AM just do the proper staging before passing the buffer back to netmod.

hzhou replied:
I think `MPIDIG_` prefix is appropriate. I agree with you the concern of mixing GPU code with AM code, and GPU fallback and AM are independent components. Ideally, the `generic` folder should be renamed to `am` folder. How about make a `staging` or `gpu` folder and put your code there? The naming is tricky. I like to use `gpu` as it is short and we all know what this is about. Anyway, let's get the structure right first and wory about the naming later.

gcongiu replied:
I moved the gpu code to `src/mpid/ch4/gpu`, although I think the best place is `src/mpid/ch4/generic`; my understanding is that we don't want to treat the GPU code as multimod netmod module as XPMEM is for shared memory. Perhaps we can group `am` and `gpu` (and future generic modules) into `generic` and use the naming scheme you suggested earlier (i.e. `MPIDIG_am_xxx` and `MPIDIG_gpu_xxx`).

gcongiu commented:
test:jenkins/ch4/ofi

gcongiu commented:
test:jenkins/ch4/ofi

gcongiu commented:
test:jenkins/ch4/ofi

gcongiu commented:
test:jenkins/ch4/ofi

gcongiu commented:
test:jenkins/ch4/ofi

gcongiu commented:
test:jenkins/ch4/ofi

gcongiu commented:
test:jenkins/ch4/ofi

hzhou commented on src/mpid/ch4/generic/gpu/mpidig_gpu_recv.c:
```diff
@@ -0,0 +1,81 @@
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDIG_GPU_RECV);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDIG_GPU_RECV);
+
+    buf = MPIDI_gpu_stage_recv_buf(buf, count, datatype);
```
Instead of overwriting `buf`, it is probably better to use a separate local variable.

hzhou replied:
Probably is better to have it return `mpi_errno`.

gcongiu replied:
Probably, this was made as it is when I was invoking the function inside the netmod so to avoid changing the buffer name in other places. Now that the call is confined inside the GPU generic code I can change this to also return additional buffer information and store that inside `MPIDIG_gpu_req_t` instead of using a map.

hzhou commented on src/mpid/ch4/generic/mpidig_init.c:
```diff
@@ -18,6 +18,7 @@ int MPIDIG_init(void)
     MPIDIG_am_init();
+    MPIDIG_gpu_init();
     MPIDIU_map_create((void **) &(MPIDI_global.win_map), MPL_MEM_RMA);
     MPIDI_global.csel_root = NULL;
```
I am surprised this is here. In fact, probably the entire `MPIDIG_init` should just be in `ch4_init`. There is no relationship between `MPIDIG_am` and `MPIDIG_gpu`. They are just two "generic" components that can be independently used.

gcongiu replied:
You are right the two are independent but still why should not these be initialized in a generic init?

hzhou commented on src/mpid/ch4/include/mpidpre.h:
```diff
@@ -219,6 +227,8 @@ typedef struct {
 #endif
 
+    MPIDIG_gpu_req_t gpu;
+
```
Why do we need it here? I thought buffer is not part of the persistent request.

gcongiu replied:
I think we had this discussion already, this is common to netmod and shmmod so should be accessible regardless the path used (either direct netmod or AM)

gcongiu commented on src/include/mpiimpl.h:
```diff
@@ -23,6 +23,7 @@
 #define MPIR_EXTERN extern
 #endif
 
+#define HAVE_GPU
```
yes I will do that

gcongiu commented on src/mpid/ch4/generic/gpu/Makefile.mk:
```diff
@@ -0,0 +1,19 @@
+
+AM_CPPFLAGS += -I$(top_srcdir)/src/mpid/ch4/generic/gpu
+
+noinst_HEADERS   += src/mpid/ch4/generic/gpu/mpidig_gpu_impl.h \
```
probably

shawnccx commented on src/mpid/ch4/generic/gpu/mpidig_gpu_recv.c:
```diff
@@ -0,0 +1,81 @@
+        MPIDI_NM_mpi_recv(buf, count, datatype, rank, tag, comm, context_offset, av, status,
+                          request);
+    MPIR_ERR_CHECK(mpi_errno);
+
```
@gcongiu I'm watching on this PR also, one thing I'm not clear is that I didn't see the data movement from intermediate CPU buffer back to original GPU buffer after send/recv completes. Can you point it to me? Or it is just draft version now?

gcongiu replied:
The data movement is done when the request is freed (and the buffer unstaged/released):         https://github.com/pmodels/mpich/blob/511e80d193864adf430fbf15eedaeb7ad687ce77/src/mpid/ch4/include/mpidpost.h#L41

shawnccx replied:
I see, thanks.

gcongiu commented on src/mpid/ch4/generic/gpu/mpidig_gpu_impl.h:
```diff
@@ -0,0 +1,173 @@
+        (MPIDIG_gpu_stage_ctx_t *) MPIDIU_map_lookup(MPIDIG_gpu_global.buf_map, (uint64_t) buf);
+    void *stage_buf = (char *) buf + ctx->true_lb;
+    void *src_buf = (char *) ctx->src_buf + ctx->true_lb;
+    MPL_gpu_memcpy_h2d(src_buf, stage_buf, ctx->size);
```
`MPL_gpu_memcpy_h2d` will be replaced by `yaksa_ipack` (or some other `MPIR_xxx` wrapper interface for yaksa).

gcongiu commented on src/mpid/ch4/generic/am/mpidig_am_init.c:
```diff
@@ -56,7 +56,8 @@ int MPIDIG_am_init(void)
 
-    MPIDI_global.buf_pool = MPIDIU_create_buf_pool(MPIDIU_BUF_POOL_NUM, MPIDIU_BUF_POOL_SZ);
+    MPIDI_global.buf_pool =
+        MPIDIU_create_buf_pool(MPIDIU_BUF_POOL_NUM, MPIDIU_BUF_POOL_SZ, MPL_GPU_POINTER_HOST);
```
@yfguo should the GPU prefix be removed from `MPL_GPU_POINTER_HOST/DEV`?

shawnccx commented on src/mpid/ch4/generic/am/mpidig_am_init.c:
```diff
@@ -56,7 +56,8 @@ int MPIDIG_am_init(void)
 
-    MPIDI_global.buf_pool = MPIDIU_create_buf_pool(MPIDIU_BUF_POOL_NUM, MPIDIU_BUF_POOL_SZ);
+    MPIDI_global.buf_pool =
+        MPIDIU_create_buf_pool(MPIDIU_BUF_POOL_NUM, MPIDIU_BUF_POOL_SZ, MPIDIU_BUF_POOL_PAGEABLE);
```
Leave a comment here: 

We made an agreement on last online meeting that, generic GPU is serving as an utility layer for two parts: data movement and pipeline structure. 

Every layer, for example `netmod/ofi` or `ch4/generic/am` layer, if it misses internal capability for handling GPU, it should call utility function from GPU utility layer.

To follow that design, buffer pool registration should happen inside GPU utility layer, not here in AM.

shawnccx commented:
Leave a note here:

We made an agreement on last online meeting that, generic GPU is serving as an utility layer for two parts: data movement and pipeline structure.

Every layer, for example `netmod/ofi` or `ch4/generic/am` layer, if it misses internal capability for handling GPU, it should call utility function from GPU utility layer.

gcongiu commented:
@shawnccx I will updpate this PR removing the gpu generic code and leaving the utility functions.

gcongiu commented:
@shawnccx I have updated this PR. Since it was not too much work I also added the fallback in the OFI netmod. I still need to test it however. Also, there might be some small tuning needed, let me know in case. Overall this should be pretty close to the final code.

shawnccx commented:
I will look through this asap, thanks.

shawnccx commented on src/mpid/ch4/generic/gpu/mpidig_gpu_util.h:
```diff
@@ -0,0 +1,182 @@
+
+    MPIDI_Datatype_get_info(count, datatype, dt_contig, size, dt_ptr, true_lb);
+    if (!dt_contig)
+        size = (dt_ptr->extent * count) + (dt_ptr->true_ub - true_lb) - dt_ptr->extent;
```
I'm not very clear about datatype, here why we need take out one `extent` size?

shawnccx replied:
Also, why can I not use the size returned from `MPIDI_Datatype_get_info`? Try to understand how buf sz is calculated here.

gcongiu replied:
The datatype extent might have been resized (i.e. the extent has not be modified using `MPI_Type_create_resized`), therefore we have to add the `true_extent` to and remove the extent to make the size right. If the datatype is not resized then `true_lb = (dt_ptr->true_ub - true_lb)` is equal to `dt_ptr->extent` and you are left with `size = dt_ptr->extent * count`.

gcongiu replied:
Sorry I missed the other comment, `MPIDI_Datatype_get_info` returns the number of useful bytes not the extent of the datatype.

shawnccx replied:
Okay, it makes sense.

gcongiu replied:
BTW I will split this PR into two. Leave this as it is and then create one for GPU utility functions. Eventually this PR will be rebased on top of that.

shawnccx commented on src/mpid/ch4/netmod/ofi/ofi_recv.h:
```diff
@@ -289,6 +289,58 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_do_irecv(void *buf,
 
 
+#ifdef MPL_HAVE_GPU
+static inline int MPIDI_NM_do_gpu_recv(void *buf,
```
This function actually does more then `gpu_recv`, it supports both cpu and gpu memory. For non-gpu memory, it just checks pointer property and call CPU transfer.

How about renaming it to `MPIDI_NM_enable_gpu_recv`, to make it more readable?

shawnccx commented on src/mpid/ch4/netmod/ofi/ofi_recv.h:
```diff
@@ -289,6 +289,58 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_do_irecv(void *buf,
 
 
+#ifdef MPL_HAVE_GPU
+static inline int MPIDI_NM_do_gpu_recv(void *buf,
```
This function is very similar to `MPIDI_NM_do_gpu_irecv`. Both of them are doing pointer check, staging recv buffer and calling internode CPU transfer.

In future, if `imrecv` is supported, we will add another `MPIDI_NM_do_gpu_imrecv` with lots of duplication.

Is it possible to have one and reuse it for `NM_recv`, `NM_irecv` and `NM_imrecv` so on?

shawnccx replied:
I think it is the same case for `send`, is it possible to have only one `MPIDI_NM_do_gpu_isend`, which can be reused for `send`, `isend`, 'ssend'?

gcongiu replied:
Yes it should be possible to reuse these but adding additional flags as `NM_recv` and `NM_irecv` have different `MPIDIG_xxx` fallbacks.

gcongiu replied:
On second thought I don't think it should be too problematic to have a little bit of duplication as these function will go away eventually when libfabric adds native support for GPU memory. But if this is important to you we can change them.

shawnccx replied:
That's fine, we can keep it for now. I think `isend` and `irecv` are most important functions to support in the beginning.

gcongiu replied:
I have changed the `NM_do_gpu_xxx` function names to a more generic `NM_do_xxx`. We usually use the `do_xxx` in the name to indicate functions that are used locally.

gcongiu commented:
test:mpich/warnings

gcongiu commented:
test:mpich/warnings

gcongiu commented:
test:mpich/warnings

shawnccx commented on src/mpid/ch4/netmod/ofi/ofi_send.h:
```diff
@@ -384,6 +384,56 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_send(const void *buf, MPI_Aint count, MPI
+    MPL_gpu_query_pointer_type(buf, &type);
+    if (!MPIDI_OFI_global.can_handle_gpu_mem &&
+        MPIDIU_map_lookup(MPIDIG_gpu_global.buf_map,
+                          (uint64_t) buf) != MPIDIU_MAP_NOT_FOUND && type == MPL_GPU_POINTER_DEV) {
```
I think the condition should be 
`MPIDIU_map_lookup(MPIDIG_gpu_global.buf_map,(uint64_t) buf) == MPIDIU_MAP_NOT_FOUND`

`NOT_FOUND` means we need to allocate stage buffer for gpu pointer, move data and return its pointer.

Otherwise it means currently there is going-on ops on this gpu pointer.

Same for both send and recv cases.

shawnccx replied:
A second thought, we don't need this check `MPIDIU_map_lookup(MPIDIG_gpu_global.buf_map, (uint64_t) buf) != MPIDIU_MAP_NOT_FOUND`

Always fetch a CPU buffer for GPU buffer.

MPIDIU_map is used to remember fetched CPU buffer, and to release it correctly in request completion.

gcongiu replied:
You are right this design was when we were falling back to the generic layer and then back again into the netmod. This check should be removed now that we are handling everything inside OFI

shawnccx replied:
I will do that in my PR.

gcongiu replied:
If it's not too much work would it be possible for you to keep working on this. You can overwrite any change you want in this PR (I will assign this to you). Otherwise do you have a PR to replace this already? If so can you update the wiki page with a link to it so that we can keep track of what is going on and close this one?

shawnccx replied:
I mean `overwrite` this one when I have updates.

shawnccx replied:
After I get local cuda tests passed.

shawnccx commented:
Updated based on #4455

shawnccx commented:
test:mpich/warnings

shawnccx commented:
@gcongiu 
Can you change the status from `draft` to `open`? (since you are the original author of this PR)
Also, I think the description is not appropriate for current change, also needs to be removed.

@pavanbalaji 
This is the OFI internode fallback path PR, it also requires your review. 
The merge pipeline should be:

1. yaksa integration https://github.com/pmodels/mpich/pull/4478
2. generic gpu utility funcs https://github.com/pmodels/mpich/pull/4455
3. then this one

gcongiu commented on src/mpid/ch4/netmod/ofi/ofi_send.h:
```diff
@@ -269,6 +282,10 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_send_normal(const void *buf, MPI_Aint cou
   fn_fail:
+    if (data_sz && attr == MPL_GPU_POINTER_DEV) {
+        /* Free stage_buf (asigned to send_buf already) */
+        MPIDIG_gpu_free_stage_buf(send_buf);
```
Shouldn't you reset `MPIDIG_GPU_REQUEST(sreq, stage_buf) = NULL`?

shawnccx commented:
test:mpich/ch4/most
test:mpich/ch3/most

shawnccx commented:
See a failure in my local CUDA+MPI test to verify `send/recv`, debugging now.

shawnccx commented:
test:mpich/ch4/most
test:mpich/ch3/most

shawnccx commented:
@gcongiu @pavanbalaji 
I think this PR is ready for review now.

FYI @zhenggb72

gcongiu commented:
@shawnccx is this ready for review?

pavanbalaji commented:
Can you fix the terminology here?  "Stage buffer" does not necessarily mean "host buffer".  It could be on the device too.  You would stage the data into some temporary buffer.  For example, if the source data is noncontiguous, you might want to stage it in a contiguous buffer on the device before handing it off to the network.

shawnccx commented:
> Can you fix the terminology here? "Stage buffer" does not necessarily mean "host buffer". It could be on the device too. You would stage the data into some temporary buffer. For example, if the source data is noncontiguous, you might want to stage it in a contiguous buffer on the device before handing it off to the network.

Okay, I will use `host_buf` and `device_buf` only.

shawnccx commented:
A few feedback after discussing with @pavanbalaji 
- `MPL_gpu_query_pointer_type` return both pointer type and the actual device (device id)
- Adding a boolean flag in `MPIDI_OFI_global_t`, showing if OFI can handle GPU data internally. It should be queried in OFI init time. (Need to follow up with OFI guys to know related OFI interfaces).
- In `recv`, moving data back to GPU should not happen in `request free` hook. `Request free` means all data movement is complete. We should add an extra counter to internal `completion counter`, and move data back to GPU when counter reaches 1. (Similar to free host_buffer case in `send`)

gcongiu commented:
@shawnccx as for the first point `MPL_gpu_query_pointer_type` only returns the pointer type not the device id. The second point sounds reasonable, right now the flag in OFI is set statically but it should indeed be queried at init time. I am not sure there is an API for that in libfabric yet. The third point also sounds reasonable. If I understand it correctly, we should increment the reference counter to account for the additional unstage operation and further decrement the ref counter after unstage is done (?).

shawnccx commented:
> If I understand it correctly, we should increment the reference counter to account for the additional unstage operation and further decrement the ref counter after unstage is done (?).

Correct. @gcongiu

shawnccx commented:
> @shawnccx as for the first point `MPL_gpu_query_pointer_type` only returns the pointer type not the device id.

@pavanbalaji @gcongiu 
How about change `MPL_gpu_query_pointer_type` to `MPL_gpu_get_pointer_attributes`? By referring `cudaPointerGetAttributes`. Both type and device is are queried.

pavanbalaji commented:
> > @shawnccx as for the first point `MPL_gpu_query_pointer_type` only returns the pointer type not the device id.
> 
> @pavanbalaji @gcongiu
> How about change `MPL_gpu_query_pointer_type` to `MPL_gpu_get_pointer_attributes`? By referring `cudaPointerGetAttributes`. Both type and device is are queried.

Yeah, that'd be the right way to do it as the native GPU APIs give both at the same time anyway.

pavanbalaji commented:
> I am not sure there is an API for that in libfabric yet.

I thought someone at Intel (@wesbland ?) had told me that libfabrics added capability for querying whether it has GPU support or not.  Or am I mistaken?

gcongiu commented:
> > @shawnccx as for the first point `MPL_gpu_query_pointer_type` only returns the pointer type not the device id.
> 
> @pavanbalaji @gcongiu
> How about change `MPL_gpu_query_pointer_type` to `MPL_gpu_get_pointer_attributes`? By referring `cudaPointerGetAttributes`. Both type and device is are queried.

I have a PR for that https://github.com/pmodels/mpich/pull/4522

wesbland commented:
> > I am not sure there is an API for that in libfabric yet.
> 
> I thought someone at Intel (@wesbland ?) had told me that libfabrics added capability for querying whether it has GPU support or not. Or am I mistaken?

I don't think this was me. I haven't been following the GPU additions to libfabric that closely at the moment. Probably @shawnccx / @sagarth / @zhenggb72 .

pavanbalaji commented:
> I don't think this was me. I haven't been following the GPU additions to libfabric that closely at the moment. Probably @shawnccx / @sagarth / @zhenggb72 .

@shawnccx  @gcongiu 

The answer to this question is critical for this PR.  Otherwise, we won't know when we should stage data and when libfabrics can handle it directly.  In case libfabrics does not give this capability yet, we should at least add a CVAR as a workaround and have the user explicitly set it when she knows that the libfabrics provider does in deed provide GPU support.  It's surely silly to rely on the user to set this, but at least we can move forward on this PR that way even if libfabrics is not quite ready yet.

shawnccx commented:
> The answer to this question is critical for this PR. Otherwise, we won't know when we should stage data and when libfabrics can handle it directly. In case libfabrics does not give this capability yet, we should at least add a CVAR as a workaround and have the user explicitly set it when she knows that the libfabrics provider does in deed provide GPU support. It's surely silly to rely on the user to set this, but at least we can move forward on this PR that way even if libfabrics is not quite ready yet.

@pavanbalaji 

- OFI reply from Sean: `We're working on this, but the FI_HMEM capability bit would be used for GPU support.`

- I can add a CVAR called `MPIR_CVAR_CH4_OFI_ALLOW_GPU_FALLBACK`, which enables fallback path for GPU.

gcongiu commented:
@shawnccx from what I can see it seems it should be possible to query `FI_HMEM` support through `fi_getinfo` even now.

shawnccx commented:
> @shawnccx from what I can see it seems it should be possible to query `FI_HMEM` support through `fi_getinfo` even now.

I can add that `flag` check if you guys would like. But I'm sure if it is standard for OFI for now.

pavanbalaji commented:
> > @shawnccx from what I can see it seems it should be possible to query `FI_HMEM` support through `fi_getinfo` even now.
> 
> I can add that `flag` check if you guys would like. But I'm sure if it is standard for OFI for now.

What does "being standard" mean?  That it's not a part of a GA release?  I'm OK with relying on the very latest OFI version in MPICH, for this feature.

shawnccx commented:
> That it's not a part of a GA release? I'm OK with relying on the very latest OFI version in MPICH, for t

What he means is that the flag is defined, but lots of providers haven't enabled the check for this flag for now. Well, I can add the flag check.

gcongiu commented:
@shawnccx as long as the flag is set to zero if provider does not has support for HMEM it should be fine.

shawnccx commented:
Since `FI_HMEM` is a capability flag in `fi_info`, I will add it as a new capability MACRO in `ofi_capability_sets.h` as well as `MPIDI_OFI_capabilities_t`. And turn it off for current providers (psm2, sockets, BGQ, default, minimum). I checkout our system, didn't see sockets and psm2 providers support this flag.

gcongiu commented:
> Since `FI_HMEM` is a capability flag in `fi_info`, I will add it as a new capability MACRO in `ofi_capability_sets.h` as well as `MPIDI_OFI_capabilities_t`. And turn it off for current providers (psm2, sockets, BGQ, default, minimum). I checkout our system, didn't see sockets and psm2 providers support this flag.

I am not very familiar with libfabric, so correct me if I am wrong. `fi_getinfo` returns the capability set of the provider in the info object. We just need to initialize the info bits to 0 and if the provider does not support GPU memory it won't set the corresponding bit. If this is correct all providers, even those that do not check for `FI_HMEM` will work as they will just leave the bit to 0. Is this correct?

shawnccx commented:
> > Since `FI_HMEM` is a capability flag in `fi_info`, I will add it as a new capability MACRO in `ofi_capability_sets.h` as well as `MPIDI_OFI_capabilities_t`. And turn it off for current providers (psm2, sockets, BGQ, default, minimum). I checkout our system, didn't see sockets and psm2 providers support this flag.
> 
> I am not very familiar with libfabric, so correct me if I am wrong. `fi_getinfo` returns the capability set of the provider in the info object. We just need to initialize the info bits to 0 and if the provider does not support GPU memory it won't set the corresponding bit. If this is correct all providers, even those that do not check for `FI_HMEM` will work as they will just leave the bit to 0. Is this correct?

You're right about `fi_getinfo`. If set capability flag to be 0 in `fi_info`, provider would turn it on based on its internal support. If set capability flag to be 1 in `fi_info`, provider would return failure if not supported.

However, in mpich-ch4, we initialize OFI provider in two ways:
- If runtime capability is enabled, a few available providers are queried by using `fi_getinfo(fi_info(0))`, for example, a MACRO `MPIDI_OFI_ENABLE_HMEM` (my upcoming change) is set 1 to if there is a provider which turns on `FI_HMEM` in `capabilities`.

- If runtime capability is disabled, then `MPIDI_OFI_ENABLE_HMEM` is decided by pre-defined per-provider value in `ofi_capabilities.h`. Currently, I think for psm2, sockets, rxm and bgq, this flag is all 0 from `fi_getinfo` (I set those per-provider value in upcomig change also), so in that case, MPIDI_OFI_ENABLE_HMEM is 0.

shawnccx commented:
Talked with @pavanbalaji 
We don't need a CVAR if OFI can tell GPU support or not.
I added an OFI GPU support capability flag in upcoming change.

pavanbalaji commented:
> Talked with @pavanbalaji
> We don't need a CVAR if OFI can tell GPU support or not.
> I added an OFI GPU support capability flag in upcoming change.

@shawnccx Is that not in the libfabric git repo yet?  If not, how will we test it?  Is there any provider that supports GPU Direct RDMA at this point?

shawnccx commented:
> > Talked with @pavanbalaji
> > We don't need a CVAR if OFI can tell GPU support or not.
> > I added an OFI GPU support capability flag in upcoming change.
> 
> @shawnccx Is that not in the libfabric git repo yet? If not, how will we test it? Is there any provider that supports GPU Direct RDMA at this point?

@pavanbalaji 
`FI_HMEM` cap flag is already in libfabric repo.
However, the providers in mpich (psm2, sockets, ...) don't support GPU RDMA at this point.
But the capability query code is the same, only checking the cap flag.
Only difference is for now, I set all GPU cap flag to be false for all mpich providers, and we need to turn it to true if some provider in future supports it.

Also in runtime mode, we check this cap flag dynamic based on the provider selected, this is no default value there.

pavanbalaji commented:
> @pavanbalaji
> `FI_HMEM` cap flag is already in libfabric repo.
> However, the providers in mpich (psm2, sockets, ...) don't support GPU RDMA at this point.

I assume, when you say "providers in mpich", you mean the providers that we have static fast paths for.  Having a fast-path is different from "supporting".  MPICH can support any provider at runtime, that meets the minimum feature requirement.

> Also in runtime mode, we check this cap flag dynamicly based on the provider selected, this is no default value there.

How will we test this if there's no provider that supports GPU Direct RDMA?

shawnccx commented:
> > @pavanbalaji
> > `FI_HMEM` cap flag is already in libfabric repo.
> > However, the providers in mpich (psm2, sockets, ...) don't support GPU RDMA at this point.
> 
> I assume, when you say "providers in mpich", you mean the providers that we have static fast paths for. Having a fast-path is different from "supporting". MPICH can support any provider at runtime, that meets the minimum feature requirement.
> 
> > Also in runtime mode, we check this cap flag dynamicly based on the provider selected, this is no default value there.
> 
> How will we test this if there's no provider that supports GPU Direct RDMA?

@pavanbalaji 
This discussion goes back to original question again.
So the question is, Sean Hefty from libfabric told that `FI_HMEM` is cap flag to check for GPU support, and this cap fag is already in libfabric.
Current situation is: no provider supports GPU RDMA for now and `FI_HEM` is turned off for all.

So for now for this PR, do we want to rely on this flag check (we know it will always return false for sure), or relying on a CVAR from user to take fallback path?

shawnccx commented:
Summarize a quick call:
Fast paths turn off `OFI_RDMA` for now.
Runtime check determines it by querying `FI_HEM` flag from provider.

We expect Cray provider will provider `OFI RDMA` and be able to be test for that path.

zhenggb72 commented:
My understanding is that we only need to check FI_HMEM. If info->caps & FI_HMEM is true, MPICH assumes that it can basically call fi_tsend, etc functions with GPU pointers directly. i.e. libfabric can handle both the local communication (shm provider, but MPICH does not have to use it), and remote communication with GPU buffers. We don't have a provider to test this capability for now, but the provider is supposed to work if it claims to support FI_HMEM.  It is virtually equivalent to if we build MPICH as netmod-only mode.

shawnccx commented:
@pavanbalaji @gcongiu 
I've update this PR include:
- `MPIDI_OFI_ENABLE_HMEM` is added to indicate `fi_hmem` support from provider
- Moving data back to GPU and freeing stage buffer are triggered when request counter hits zero, before `free_request` hook.

Please take a second look and let me know anything to fix.

shawnccx commented:
test:mpich/ch4/most

pavanbalaji commented on src/mpid/ch4/generic/gpu/mpidig_gpu_utils.h:
```diff
@@ -5,8 +5,6 @@
 #ifndef MPIDIG_GPU_UTIL_H_INCLUDED
 #define MPIDIG_GPU_UTIL_H_INCLUDED
 
-#include "ch4_impl.h"
```
Why is this removed?  We need `mpiimpl.h` or something here.  Also, this seems unrelated to this PR.

shawnccx replied:
The functions in this file are from `ch4_impl.h` anymore, this is a leftover remove from https://github.com/pmodels/mpich/pull/4528. Leaving include `ch4_impl.h` here would cause recursive header inclusion issues. Because request completion in `ch4_impl.h` calls functions here.

pavanbalaji replied:
Can you separate out that patch with the appropriate commit message explaining it?

shawnccx replied:
Will do.

pavanbalaji commented on src/mpid/ch4/netmod/ofi/ofi_events.c:
```diff
@@ -309,6 +309,13 @@ int MPIDI_OFI_send_event(struct fi_cq_tagged_entry *wc, MPIR_Request * sreq, int
             MPL_free(MPIDI_OFI_REQUEST(sreq, noncontig.nopack));
 
+        /* host buf != NULL means this request takes fall-back path
+         * for GPU data. */
```
The term "fallback" is overloaded.  We use that when we fallback to AM-based implementations too.  Maybe call it "fallback to host-buffer staging"?

pavanbalaji commented on src/mpid/ch4/netmod/ofi/ofi_recv.h:
```diff
@@ -104,12 +104,17 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_do_irecv(void *buf,
     uint64_t match_bits, mask_bits;
     MPIR_Context_id_t context_id = comm->recvcontext_id + context_offset;
-    size_t data_sz;
+    /* Add initial value to for check in fn_fail. */
```
I think this comment can be removed.

pavanbalaji commented on src/mpid/ch4/netmod/ofi/ofi_recv.h:
```diff
@@ -104,12 +104,17 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_do_irecv(void *buf,
     MPIR_Datatype *dt_ptr;
     struct fi_msg_tagged msg;
-    char *recv_buf;
+    /* Set initial value for compiler uninitialization complain */
```
This comment can be removed.  Also, if this compiler warning was occurring before this PR, it should be a separate patch.

pavanbalaji commented on src/mpid/ch4/netmod/ofi/ofi_recv.h:
```diff
@@ -104,12 +104,17 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_do_irecv(void *buf,
-    char *recv_buf;
+    /* Set initial value for compiler uninitialization complain */
+    char *recv_buf = NULL;
+    /* Give pointer type an initial value, in case fail before allocating stage host buffer. */
```
Comment can be removed.

pavanbalaji commented on src/mpid/ch4/netmod/ofi/ofi_recv.h:
```diff
@@ -134,6 +139,22 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_do_irecv(void *buf,
+            /* Fallback path, create a CONTIGUOUS host buf to receive data.
+             * Move data from host buf to gpu buf when receive completes. */
+            void *host_buf = NULL;
+            MPL_gpu_malloc_host(&host_buf, data_sz);
```
This is a very large memory allocation.  Shouldn't it be pipelined?

shawnccx replied:
But I think this path will be replaced in future by pipeline work. At this point, we don't have pipeline infrastructure, so here is to make sure GPU data can be transferred.

pavanbalaji replied:
No, it'll be replaced by the iQueue infrastructure that'll maintain a pool of buffers for you.  But it'll not add pipelining.  You'll need to do pipelining here.  What I was suggesting was that you can malloc/free every time, but you still need to maintain state about how many bytes have been transferred and how many are left.  Once iQueue is ready, you can replace that malloc/free with getting buffers from the iQueue pool.

shawnccx replied:
I will add `FIXME` for future work, large message should go through chunked pipeline and use fi_read.

shawnccx commented:
@raffenet Can you help upgrade the libfabric on jenkins system.
It is too old to have `FI_HMEM` flag.

For example:
https://jenkins-pmrs.cels.anl.gov/job/mpich-review-ch4-ofi/1165/compiler=gnu,jenkins_configure=external,label=centos64_review/artifact/config.log

It uses `--with-libfabric=/usr`

shawnccx commented:
test:mpich/ch4/most

shawnccx commented:
@pavanbalaji @gcongiu 
Please take a look of changes again.
I have added new changes:
- A check in `autoconf` is added due to `FI_HMEM` is not defined in older versions of libfabric, for older version, I skip `FI_HMEM` check and set `MPIDI_OFI_ENABLE_HMEM` to be 0.
- `FIXME` added to comment that future work is required to apply pipeline for large message transfer.
- Some code comments fixed.

hzhou commented on src/mpid/ch4/netmod/ofi/ofi_recv.h:
```diff
@@ -104,12 +104,13 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_do_irecv(void *buf,
     uint64_t match_bits, mask_bits;
     MPIR_Context_id_t context_id = comm->recvcontext_id + context_offset;
-    size_t data_sz;
+    size_t data_sz = 0;
```
Why do we initialize to 0 here? If there is a warning on uninitialized usage, we need set values in each code path. Otherwise, how do we know that this `0` works for each code path?

shawnccx replied:
In case you fail at line #124, data_sz is still uninitialized over there, but in `fn_fail`, data_sz is used for condition check.

shawnccx replied:
Also, `data_sz` is overwritten at line 133. Compiler just complains that it is possible for `fn_fail` check before line 133.

hzhou replied:
I see. How about use a flag so we don't have to repeat the condition check in `fn_fail`? When you repeat the condition check, whenever we update the condition logic, we need modify it twice, that's an unnecessary maintenance overhead.

shawnccx replied:
So you wanna a flag showing there is allocated CPU staging buffer, right?

hzhou replied:
Yes. That way it is explicit.

shawnccx replied:
I can add that, same for `ofi_send.h`.

shawnccx commented:
test:mpich/ch4/most

shawnccx commented:
@pavanbalaji I've update the `FI_HEM` capability flag part based on discussion we had.
@hzhou stage buffer allocation flag is introduced to avoid duplicated condition check.

Feel free to leave any comments.

shawnccx commented:
`ch4` tests pass, rebase again.
